File: src/GameLogic.java
import common.Block;
import common.IBlock;
import common.Tile;
import mob.Mob;
import mob.MobPlayer;
import weapons.GrappleGun;

import javax.swing.*;
import java.awt.*;
import java.awt.image.BufferedImage;
import java.net.URL;
import java.util.ArrayList;

import static common.Block.Type.*;


public class GameLogic implements Runnable {

    DisplayPanel j;

    int cycle = 0;
    static int currentLevel;
    final int maxLevel = 5;
    static ArrayList<Level> levelArray;
    static Mob character;
    boolean musicOn = true;

    boolean upButtonPressed = false;
    boolean leftButtonPressed = false;
    boolean rightButtonPressed = false;
    boolean downButtonPressed = false;
    boolean buttonReleased = false;

    static int mouseX;
    static int mouseY;
    static boolean mousePressed = false;
    int displacement;


    //Image initialization
    BufferedImage window, background0, background1, characterImage, inputList;
    static BufferedImage[] blockImages = {
            ResourceLoader.INSTANCE.getImage(new Block(0, 0, 0, 0, PLAIN).getImageString()),
            ResourceLoader.INSTANCE.getImage(new Block(0, 0, 0, 0, DEATH).getImageString()),
            ResourceLoader.INSTANCE.getImage(new Block(0, 0, 0, 0, SLOW).getImageString()),
            ResourceLoader.INSTANCE.getImage(new Block(0, 0, 0, 0, PLATFORM).getImageString()),
            ResourceLoader.INSTANCE.getImage(new Block(0, 0, 0, 0, PORTAL).getImageString()),
    };
    static BufferedImage[] tileImages = {
            ResourceLoader.INSTANCE.getImage("Door01.png"),
            ResourceLoader.INSTANCE.getImage("GrassBlock01.png"),
            ResourceLoader.INSTANCE.getImage("CornerBlockTopRight.png"),
            ResourceLoader.INSTANCE.getImage("CornerBlockTopLeft.png"),
            ResourceLoader.INSTANCE.getImage("Horizontal.png"),
            ResourceLoader.INSTANCE.getImage("CornerBlockBottomRight.png"),
            ResourceLoader.INSTANCE.getImage("CornerBlockBottomLeft.png"),
            ResourceLoader.INSTANCE.getImage("BottomHorizontal.png"),
    };
    static BufferedImage[] BackgroundImages = {
            ResourceLoader.INSTANCE.getImage("sword-and-sworcery.png"),
            ResourceLoader.INSTANCE.getImage("windows_xp_bliss-wide.jpg"),
            ResourceLoader.INSTANCE.getImage("whiteBackground.png"),
    };
    //	//sound initialization
//		AudioInputStream audioIn;
//		Clip clip;
    URL soundtrack1, soundtrack2;
    URL currentSoundtrack;

    GameLogic() {
        j = new DisplayPanel();
        j.setPreferredSize(new Dimension(900, 600));
        inputList = ResourceLoader.INSTANCE.getImage("InputList2.png");
        soundtrack1 = ResourceLoader.INSTANCE.getSoundUrl("Grapple - Main Menu Theme - 5-17-14, 10.57 PM.wav");
        soundtrack2 = ResourceLoader.INSTANCE.getSoundUrl("Grapple - City (Level 01) - 5-27-14, 2.52 AM.wav");

        currentLevel = 0;
        levelArray = new ArrayList<>();
        levelArray.add(new Level(0, new Tile(0, 0, "whiteBackground.png"), soundtrack1));
        levelArray.add(new Level(1, new Tile(0, 0, "windows_xp_bliss-wide.jpg"), soundtrack2));
        levelArray.add(new Level(2, new Tile(0, 0, "whiteBackground.png"), soundtrack1));
        levelArray.add(new Level(3, new Tile(0, 0, "windows_xp_bliss-wide.jpg"), soundtrack1));
        levelArray.add(new Level(4, new Tile(0, 0, "sword-and-sworcery.png"), soundtrack1));
        levelArray.add(new Level(5, new Tile(0, 0, "sword-and-sworcery.png"), soundtrack1));

        character = new MobPlayer(5, 50, 50, 200);
        character.getWeaponArray().add(new GrappleGun());

//		fillClip();
//		currentSoundtrack = levelArray.get(currentLevel).getCurrentSoundtrack();
//		//runSoundFile(currentSoundtrack, 10);

    }

    public void fillClip() {
//		try{
//			clip = AudioSystem.getClip();
//		}
//	    catch (LineUnavailableException e) {
//	         e.printStackTrace();
//	    }
    }

    public void runSoundFile(URL soundUrl, int loops) {
//		try {
//	         // Open an audio input stream.
//	         audioIn = AudioSystem.getAudioInputStream(soundUrl);
//	         // Get a sound clip resource.
//	         //clip = AudioSystem.getClip();
//	         // Open audio clip and load samples from the audio input stream.
//	         clip.open(audioIn);
//	         if(musicOn)clip.loop(loops);
//
//	      } catch (UnsupportedAudioFileException e) {
//	         e.printStackTrace();
//	      } catch (IOException e) {
//	         e.printStackTrace();
//	      } catch (LineUnavailableException e) {
//	         e.printStackTrace();
//	      }
    }

    public void run() {

        character.setXacc(0);
        if (Main.getLevelChanged()) {
            changeLevel();
        }
        Main.setLevelChanged(false);

        processInput();

        Level level = levelArray.get(currentLevel);
        Tile background = level.getBackground();
        if (character.getX() < 250 && character.getXspeed() < 0 &&
                background.getX() < 0) {

            character.setXlocked(true);
            displacement = character.getXspeed();
            if (background.getX() - displacement > 0)
                displacement = (background.getX() - displacement);
            background.setX(background.getX() - displacement);
            if (background.getX() > 0) {
                displacement -= background.getX();
            }
            for (IBlock b : level.getLevelBlocks())
                b.setX(b.getX() - displacement);
            for (Tile b : level.getLevelTiles())
                b.setX(b.getX() - displacement);


        }
        boolean b1 = background.getX() + level.getLevelWidth() * level.getBlockSize() > j.getWidth();
        if (character.getX() > j.getWidth() - 250
                && character.getXspeed() > 0
                && b1) {

            character.setXlocked(true);
            displacement = character.getXspeed();
            if (level.getLevelWidth() * level.getBlockSize() - displacement < j.getWidth()) {
                displacement = level.getLevelWidth() * level.getBlockSize() - displacement - j.getWidth();
            }
            for (IBlock b : level.getLevelBlocks()) {
                b.setX(b.getX() - character.getXspeed());

            }
            for (Tile b : level.getLevelTiles()) {
                b.setX(b.getX() - character.getXspeed());
            }
            background.setX(background.getX() - character.getXspeed());
        }

        //untested start
        boolean b2 = background.getY() + level.getLevelHeight() * level.getBlockSize() > j.getHeight();
        if (character.getY() > j.getHeight() - 250
                && character.getYspeed() > 0
                && b2) {

            character.setYlocked(true);
            displacement = character.getYspeed();
            if (level.getLevelHeight() * level.getBlockSize() - displacement < j.getHeight()) {
                displacement = level.getLevelHeight() * level.getBlockSize() - displacement - j.getHeight();
            }
            for (IBlock b : level.getLevelBlocks()) {
                b.setY(b.getY() - character.getYspeed());

            }
            for (Tile b : level.getLevelTiles()) {
                b.setY(b.getY() - character.getYspeed());
            }
            background.setY(background.getY() - character.getYspeed());
        }

        //untested end

        character.tetherMove();
        character.setXlocked(false);
        character.setYlocked(false);
        character.checkWindowBoundaries(j.getWidth(), j.getHeight());
        character.checkBlockBoundaries(level.getLevelBlocks());

        for (Mob c : level.getLevelMobs()) {
            c.pattern();
            c.checkWindowBoundaries(j.getWidth(), j.getHeight());
            c.checkBlockBoundaries(level.getLevelBlocks());
        }

        j.repaint();
        cycle++;
        //System.out.println(cycle);
        try {
            Thread.sleep(29);
        } catch (InterruptedException e) {
            System.out.println("Thread was interrupted.");
        }

    }

    public void changeLevel() {
//		clip.stop();
//		clip.close();
        currentLevel += 1;
        //DisplayFrame.levelChanged = true;
        if (currentLevel > maxLevel)
            currentLevel = 0;
        currentSoundtrack = levelArray.get(currentLevel).getCurrentSoundtrack();
        //runSoundFile(currentSoundtrack, 10);
    }

    public enum gamestate {main_screen, title_screen, gameplay_screen}

    public static gamestate current_gamestate = gamestate.main_screen;

    public void processInput() {
        if (upButtonPressed) character.jump();
        if (leftButtonPressed) character.setXacc(-2);
        if (rightButtonPressed) character.setXacc(2);
        if (downButtonPressed) character.setYspeed(character.getYspeed() + 2);
    }

    public JPanel getJPanel() {
        return j;
    }

}

File: src/GameLogic.kt
import common.Block
import common.Tile
import mob.Mob
import mob.MobPlayer
import weapons.GrappleGun
import javax.swing.*
import java.awt.*
import java.awt.image.BufferedImage
import java.net.URL
import java.util.ArrayList
import common.Block.Type.*
import common.IBlock

class GameLogic constructor() : Runnable {
    var j: DisplayPanel = DisplayPanel()
    var cycle = 0
    private val maxLevel = 5
    var musicOn = true
    var upButtonPressed = false
    var leftButtonPressed = false
    var rightButtonPressed = false
    var downButtonPressed = false
    var displacement: Int = 0
    var inputList: BufferedImage
    //	//sound initialization
    //		AudioInputStream audioIn;
    //		Clip clip;
    var soundtrack1: URL
    var soundtrack2: URL

    //    internal var currentSoundtrack: URL
    init {
        j.preferredSize = Dimension(900, 600)
        inputList = ResourceLoader.getImage("InputList2.png")
        soundtrack1 = ResourceLoader.getSoundUrl("Grapple - Main Menu Theme - 5-17-14, 10.57 PM.wav")
        soundtrack2 = ResourceLoader.getSoundUrl("Grapple - City (Level 01) - 5-27-14, 2.52 AM.wav")
        currentLevel = 0

        character = MobPlayer(5, 50, 50, 200)
        character.weaponArray.add(GrappleGun())
        //		fillClip();
        //		currentSoundtrack = levelArray.get(currentLevel).getCurrentSoundtrack();
        //		//runSoundFile(currentSoundtrack, 10);
    }

    fun fillClip() {
        //		try{
        //			clip = AudioSystem.getClip();
        //		}
        //	    catch (LineUnavailableException e) {
        //	         e.printStackTrace();
        //	    }
    }

    fun runSoundFile(soundUrl: URL, loops: Int) {
        //		try {
        //	         // Open an audio input stream.
        //	         audioIn = AudioSystem.getAudioInputStream(soundUrl);
        //	         // Get a sound clip resource.
        //	         //clip = AudioSystem.getClip();
        //	         // Open audio clip and load samples from the audio input stream.
        //	         clip.open(audioIn);
        //	         if(musicOn)clip.loop(loops);
        //
        //	      } catch (UnsupportedAudioFileException e) {
        //	         e.printStackTrace();
        //	      } catch (IOException e) {
        //	         e.printStackTrace();
        //	      } catch (LineUnavailableException e) {
        //	         e.printStackTrace();
        //	      }
    }

    override fun run() {
        character.xacc = 0
        if (Main.levelChanged) {
            changeLevel()
        }
        Main.levelChanged = false

        processInput()
        val level = levelArray[currentLevel]
        val background = level.background
        if (character.x < 250 && character.xspeed < 0 &&
                background.x < 0) {
            character.xlocked = true
            displacement = character.xspeed
            if (background.x - displacement > 0)
                displacement = background.x - displacement
            background.x = background.x - displacement
            if (background.x > 0) {
                displacement -= background.x
            }
            level.levelTiles.forEach { b -> b.x = b.x - displacement }
        }
        val b1 = background.x + level.levelWidth * level.blockSize > j.width
        if (character.x > j.width - 250
                && character.xspeed > 0
                && b1) {
            character.xlocked = true
            displacement = character.xspeed
            if (level.levelWidth * level.blockSize - displacement < j.width) {
                displacement = level.levelWidth * level.blockSize - displacement - j.width
            }
            level.levelTiles.forEach { it.x = it.x - character.xspeed }
            background.x = background.x - character.xspeed
        }
        //untested start
        val b2 = background.y + level.levelHeight * level.blockSize > j.height
        if (character.y > j.height - 250
                && character.yspeed > 0
                && b2) {
            character.ylocked = true
            displacement = character.yspeed
            if (level.levelHeight * level.blockSize - displacement < j.height) {
                displacement = level.levelHeight * level.blockSize - displacement - j.height
            }
            level.levelTiles.forEach { b -> b.y = b.y - character.yspeed }
            background.y = background.y - character.yspeed
        }
        //untested end
        character.run {
            tetherMove()
            xlocked = false
            ylocked = false
            checkWindowBoundaries(j.width, j.height)
            checkBlockBoundaries(level.levelTiles.filterIsInstance(IBlock::class.java))
        }

        level.levelMobs.forEach {
            it.pattern()
            it.checkWindowBoundaries(j.width, j.height)
            it.checkBlockBoundaries(level.levelTiles.filterIsInstance(IBlock::class.java))
        }

        j.repaint()
        cycle++
        //System.out.println(cycle);
        try {
            Thread.sleep(29)
        } catch (e: InterruptedException) {
            println("Thread was interrupted.")
        }
    }

    fun changeLevel() {
        //		clip.stop();
        //		clip.close();
        currentLevel += 1
        //DisplayFrame.levelChanged = true;
        if (currentLevel > maxLevel)
            currentLevel = 0
//        currentSoundtrack = levelArray[currentLevel].currentSoundtrack
        //runSoundFile(currentSoundtrack, 10);
    }

    enum class Gamestate {
        MAIN_SCREEN, TITLE_SCREEN, GAMEPLAY_SCREEN
    }

    private fun processInput() {
        if (upButtonPressed) character.jump()
        if (leftButtonPressed) character.xacc = -2
        if (rightButtonPressed) character.xacc = 2
        if (downButtonPressed) character.yspeed = character.yspeed + 2
    }

    companion object {
        var currentLevel: Int = 0
        var levelArray: ArrayList<Level> = arrayListOf(
                Level(0, Tile(0, 0, "whiteBackground.png")),
                Level(1, Tile(0, 0, "windows_xp_bliss-wide.jpg")),
                Level(2, Tile(0, 0, "whiteBackground.png")),
                Level(3, Tile(0, 0, "windows_xp_bliss-wide.jpg")),
                Level(4, Tile(0, 0, "sword-and-sworcery.png")),
                Level(5, Tile(0, 0, "sword-and-sworcery.png"))
        )
        var character: Mob = MobPlayer(1, 1, 20, 20)
        var mouseX: Int = 0
        var mouseY: Int = 0
        var mousePressed = false
        var current_gamestate = Gamestate.MAIN_SCREEN
    }
}



